File: android/app/src/main/java/com/reymelin/gradientclock/widget/GradientClockWidgetProvider.java
package com.reymelin.gradientclock.widget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Build;
import android.os.Bundle;
import android.widget.RemoteViews;

import com.reymelin.gradientclock.MainActivity;
import com.reymelin.gradientclock.R;

public class GradientClockWidgetProvider extends AppWidgetProvider {

  @Override
  public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
    for (int appWidgetId : appWidgetIds) {
      updateOne(context, appWidgetManager, appWidgetId);
    }
    WidgetUpdateScheduler.ensureScheduled(context);
  }

  @Override
  public void onAppWidgetOptionsChanged(Context context, AppWidgetManager appWidgetManager, int appWidgetId, Bundle newOptions) {
    super.onAppWidgetOptionsChanged(context, appWidgetManager, appWidgetId, newOptions);
    updateOne(context, appWidgetManager, appWidgetId);
  }

  @Override
  public void onEnabled(Context context) {
    super.onEnabled(context);
    WidgetUpdateScheduler.ensureScheduled(context);
  }

  @Override
  public void onDisabled(Context context) {
    super.onDisabled(context);
    WidgetUpdateScheduler.cancel(context);
  }

  public static void updateAll(Context context) {
    AppWidgetManager mgr = AppWidgetManager.getInstance(context);
    int[] ids = mgr.getAppWidgetIds(new ComponentName(context, GradientClockWidgetProvider.class));
    for (int id : ids) updateOne(context, mgr, id);
  }

  private static void updateOne(Context context, AppWidgetManager mgr, int appWidgetId) {
    RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.gradient_clock_widget);

    Bundle options = mgr.getAppWidgetOptions(appWidgetId);
    int minW = options.getInt(AppWidgetManager.OPTION_APPWIDGET_MIN_WIDTH);
    int minH = options.getInt(AppWidgetManager.OPTION_APPWIDGET_MIN_HEIGHT);

    // Convert dp -> px
    float density = context.getResources().getDisplayMetrics().density;
    int pxW = (int) (minW * density);
    int pxH = (int) (minH * density);

    if (pxW == 0 || pxH == 0) {
        android.util.Log.d("Widget", "Still waiting for size, delaying update");
        return;
    }

    Bitmap bmp = GradientRenderer.renderWidgetBitmap(context, pxW, pxH);
    views.setImageViewBitmap(R.id.widget_bg, bmp);

    Intent intent = new Intent(context, MainActivity.class);
    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);

    int flags = 0;
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) flags = PendingIntent.FLAG_IMMUTABLE;

    PendingIntent pi = PendingIntent.getActivity(context, 0, intent, flags);
    views.setOnClickPendingIntent(R.id.widget_bg, pi);

    mgr.updateAppWidget(appWidgetId, views);
  }
}

File: android/app/src/main/java/com/reymelin/gradientclock/widget/GradientClockWidgetProvider.kt
package com.reymelin.gradientclock.widget

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.os.Build
import android.widget.RemoteViews
import com.reymelin.gradientclock.MainActivity
import com.reymelin.gradientclock.R

class GradientClockWidgetProvider : AppWidgetProvider() {

  override fun onUpdate(
    context: Context,
    appWidgetManager: AppWidgetManager,
    appWidgetIds: IntArray
  ) {
    for (appWidgetId in appWidgetIds) {
      updateOne(context, appWidgetManager, appWidgetId)
    }
    WidgetUpdateScheduler.ensureScheduled(context)
  }

  override fun onEnabled(context: Context) {
    super.onEnabled(context)
    WidgetUpdateScheduler.ensureScheduled(context)
  }

  override fun onDisabled(context: Context) {
    super.onDisabled(context)
    WidgetUpdateScheduler.cancel(context)
  }

  companion object {
    fun updateAll(context: Context) {
      val mgr = AppWidgetManager.getInstance(context)
      val ids = mgr.getAppWidgetIds(
        ComponentName(context, GradientClockWidgetProvider::class.java)
      )
      for (id in ids) updateOne(context, mgr, id)
    }

    private fun updateOne(
      context: Context,
      appWidgetManager: AppWidgetManager,
      appWidgetId: Int
    ) {
      val views = RemoteViews(context.packageName, R.layout.gradient_clock_widget)

      // Background gradient bitmap + current time
      val bmp = GradientRenderer.renderGradientBitmap(context)
      views.setImageViewBitmap(R.id.widget_bg, bmp)
      views.setTextViewText(R.id.widget_time, TimeFormatter.nowHHmm())

      // Tap opens app
      val intent = Intent(context, MainActivity::class.java).apply {
        flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
      }
      val flags = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
        PendingIntent.FLAG_IMMUTABLE else 0

      val pi = PendingIntent.getActivity(context, 0, intent, flags)
      views.setOnClickPendingIntent(R.id.widget_bg, pi)
      views.setOnClickPendingIntent(R.id.widget_time, pi)

      appWidgetManager.updateAppWidget(appWidgetId, views)
    }
  }
}
